internal/protoc/compiler.go

```diff
@@ -0,0 +1,736 @@
+		if err != nil {
+			return nil, err
+		}
+		pluginFlagSets = append(pluginFlagSets, pluginFlagSet)
+	}
+	return pluginFlagSets, nil
+}
+
+func getPluginFlagSet(protoSet *file.ProtoSet, dirPath string, genPlugin settings.GenPlugin) ([]string, error) {
+	protoFlags, err := getPluginFlagSetProtoFlags(protoSet, dirPath, genPlugin)
+	if err != nil {
+		return nil, err
+	}
+	flagSet := []string{fmt.Sprintf("--%s_out=%s", genPlugin.Name, genPlugin.OutputPath.AbsPath)}
+	if len(protoFlags) > 0 {
+		flagSet = []string{fmt.Sprintf("--%s_out=%s:%s", genPlugin.Name, protoFlags, genPlugin.OutputPath.AbsPath)}
+	}
+	if genPlugin.Path != "" {
+		flagSet = append(flagSet, fmt.Sprintf("--plugin=protoc-gen-%s=%s", genPlugin.Name, genPlugin.Path))
+	}
+	return flagSet, nil
+}
+
+// the return value corresponds to CodeGeneratorRequest.Parameter
+// https://github.com/golang/protobuf/blob/b4deda0973fb4c70b50d226b1af49f3da59f5265/protoc-gen-go/plugin/plugin.pb.go#L103
+// this function basically just sets the Mfile=package values for go and gogo plugins
+func getPluginFlagSetProtoFlags(protoSet *file.ProtoSet, dirPath string, genPlugin settings.GenPlugin) (string, error) {
+	// the type just denotes what Well-Known Type map to use from the wkt package
+	// if not go or gogo, we don't have any special automatic handling, so just return what we have
+	if !genPlugin.Type.IsGo() && !genPlugin.Type.IsGogo() {
+		return genPlugin.Flags, nil
+	}
+	if genPlugin.Type.IsGo() && genPlugin.Type.IsGogo() {
+		return "", fmt.Errorf("internal error: plugin %s is both a go and gogo plugin", genPlugin.Name)
+	}
+	var goFlags []string
+	if genPlugin.Flags != "" {
+		goFlags = append(goFlags, genPlugin.Flags)
+	}
+	genGoPluginOptions := protoSet.Config.Gen.GoPluginOptions
+	if !genGoPluginOptions.NoDefaultModifiers {
+		modifiers := make(map[string]string)
+		for subDirPath, protoFiles := range protoSet.DirPathToFiles {
+			// you cannot include the files in the same package in the Mfile=package map
+			// or otherwise protoc-gen-go, protoc-gen-gogo, etc freak out and put
+			// these packages in as imports
+			if subDirPath != dirPath {
+				for _, protoFile := range protoFiles {
+					path, err := filepath.Rel(protoSet.Config.DirPath, protoFile.Path)
+					if err != nil {
+						// TODO: best effort, maybe error
+						path = protoFile.Path
+					}
+					// TODO: if relative path in OutputPath.RelPath jumps out of import path context, this will be wrong
+					modifiers[path] = filepath.Clean(filepath.Join(genGoPluginOptions.ImportPath, genPlugin.OutputPath.RelPath, filepath.Dir(path)))
+				}
+			}
+		}
+		for key, value := range modifiers {
+			goFlags = append(goFlags, fmt.Sprintf("M%s=%s", key, value))
+		}
+		if protoSet.Config.Compile.IncludeWellKnownTypes {
+			// one of these two must be true, we validate this above
+			if genPlugin.Type.IsGo() {
+				modifiers = wkt.FilenameToGoModifierMap
+			}
+			if genPlugin.Type.IsGogo() {
```

Comment:
Although unnecessary, could we do an `else if` here? We could remove the comment since it wouldn't really need any additional explanation.

Reply:
I'll do it, but note above I error out if both are set.